### PR TITLE
UI: Internationalization support

### LINKED ISSUE #1070: Multilanguage support
Before openpilot is 1.0 software, it needs support for multiple languages. We'd need a nice generic way to do this.

I think this will have to be written at first in house, but it should be easy after that to add more languages with pull requests. We should look into what dragonpilot is doing and take what's good.

### PR DESCRIPTION
**Pull request by sshane**
- [x] Wrap UI strings
- [x] UI for selecting languages (with or without restarting)
- [x] Dynamic database of languages, to add a new one you should just need to add ["Spanish", "es"] for example to a list somewhere
- [x] ~~Fix ordering of the language selector~~

### CODE REVIEW

**Comment by pd0wm** [comment]
Cool! Are these xml files a widely accepter standard just something Qt specific? Is there a nice editor for them?

**Comment by ShaneSmiskol** [comment]
I'm not sure how widely accepted they are, but yes there's a GUI tool called Linguist installed with QT. Wrap a new string, run that bash script and it will show up! There's also another unofficial tool that automatically uploads your strings to Google Translate and fetches the translations, but I haven't tried it yet https://github.com/gnomezgrave/qt-auto-translator

**Comment by maxime-desroches** [comment]
DRIVES  ->  UTILISATIONS

HOURS  ->  HEURES

ALL TIME  ->  DEPUIS TOUJOURS

PAST WEEK  ->  CETTE SEMAINE

Become a comma prime member at my.comma.ai and get premium features!  ->  Devenez membre comma prime sur my.comma.ai et bénéficiez de fonctionnalités premium&#xa0;!

✓ DEVELOPER PERKS  ->  ✓ AVANTAGES POUR LES DÉVELOPPEURS

COMMA POINTS  ->  POINTS COMMA

Scan with comma connect!  ->  Scannez-moi avec comma connect!

**Comment by pd0wm** [comment]
Let's not worry about the actual translations for now. Let's get the infrastructure in place and then add new languages through separate PRs.

**Comment by spektor56** [comment]
does this mean you are going to change the training screens from PNG's to actual Text + images?

File: selfdrive/ui/qt/offroad/settings.cc [hunk 1]
```diff
@@ -133,6 +146,20 @@ DevicePanel::DevicePanel(SettingsWindow *parent) : ListWidget(parent) {
...
+  auto translateBtn = new ButtonControl(tr("Change Language"), tr("CHANGE"), "");
+  connect(translateBtn, &ButtonControl::clicked, [=]() {
+    QMap<QString, QString> langs = getSupportedLanguages();
+    QString selection = MultiOptionDialog::getSelection(tr("Select a language"), langs.keys(), this);
+    qDebug() << "Selected:" << selection << langs[selection];
+
+    if (!selection.isEmpty()) {
+      Params().put("LanguageSetting", langs[selection].toStdString());
+      qApp->quit();
+      QProcess::startDetached(qApp->arguments()[0], qApp->arguments());
```
[the remark this answers is not in the document]
Comment (sshane) [reply selfdrive/ui/qt/offroad/settings.cc:158]: No it doesn't, one of the problems I was thinking about. We can just kill the app, but then it requires running UI through manager to restart properly and I think the delay might not be great.

[the remark this answers is not in the document]
Comment (sshane) [reply selfdrive/ui/qt/offroad/settings.cc:158]: We can try to update the watchdog file with the new pid every time UI starts

[the remark this answers is not in the document]
Comment (pd0wm) [reply selfdrive/ui/qt/offroad/settings.cc:158]: we can add an argument to the ui to override which pid to use for the watchdog? Alternatively you can try writing 0 into the watchdog file, that should also trigger a restart on the next manager iteration.

[the remark this answers is not in the document]
Comment (sshane) [reply selfdrive/ui/qt/offroad/settings.cc]: @adeebshihadeh suggested looking at the proc's exitcode.

**Comment by sshane** [comment]
@adeebshihadeh can you check out the changes in process.py?

File: selfdrive/ui/qt/offroad/settings.cc [hunk 2]
```diff
@@ -133,6 +134,19 @@ DevicePanel::DevicePanel(SettingsWindow *parent) : ListWidget(parent) {
...
   }
 
+  auto translateBtn = new ButtonControl(tr("Change Language"), tr("CHANGE"), "");
+  connect(translateBtn, &ButtonControl::clicked, [=]() {
+    QMap<QString, QString> langs = getSupportedLanguages();
+    QString selection = MultiOptionDialog::getSelection(tr("Select a language"), langs.keys(), this);
+    if (!selection.isEmpty()) {
+      // put language setting, exit Qt UI, and trigger fast restart
+      Params().put("LanguageSetting", langs[selection].toStdString());
+      qApp->exit(18);
```
Comment (sshane) [inline selfdrive/ui/qt/offroad/settings.cc:144]: In case we want to filter these events out, check for 18 exit code (i18n)

**Pull request merged**